Adapter/Translator.php

```diff
@@ -0,0 +1,143 @@
+     * @param array       $parameters An array of parameters for the message
+     * @param string|null $module     The domain: Legacy Admin controller name, module name or null to use the Core. Other special value: 'pdf'.
+     * @param string|null $locale     The locale or null to use the default
+     *
+     * @throws \InvalidArgumentException If the locale contains invalid characters
+     *
+     * @return string The translated string
+     *
+     * @api
+     */
+    public function trans($id, array $parameters = array(), $domain = null, $locale = null)
+    {
+        // Very speific case (Form error)
+        if ($domain == 'form_error') {
+            return \Tools::displayError($id, false);
+        }
+        // Very specific cases (PDF)
+        if ($domain == 'pdf') {
+            // Does not support overriding the language for this adapter!
+            return \Translate::getPdfTranslation($id, (count($parameters) === 0) ? null : $parameters);
+        }
+
+        // Search for Admin case
+        $isAdmin = ((isset($this->context->controller) && $this->context->controller->controller_type == 'admin')
+                    || $this->context->get('app_entry_point', 'unknown') === 'admin');
+        if ($isAdmin && $domain == null) {
+            $domain = 'AdminTab'; // default class value for legacy Admin translation
+        }
+        if ($isAdmin) {
+            $domain = preg_replace('/(c|C)ontroller$/', '', $domain); // remove trailing 'Controller'
+
+            return \Translate::getAdminTranslation($id, $domain, (count($parameters) === 0) ? null : $parameters);
+        }
+
+        // Front / Module case ?
+        if ($domain !== null) {
+            throw new DevelopmentErrorException('Module & Front translation is not yet implemented. Please contact the Architect team.');
+        }
+
+        throw new DevelopmentErrorException('Translation without $domain key is not yet implemented in the front interface. Please contact the Architect team.');
+    }
+
+    /**
+     * Translates the given choice message by choosing a translation according to a number.
+     *
+     * @param string      $id         The message id (may also be an object that can be cast to string)
+     * @param int         $number     The number to use to find the indice of the message
+     * @param array       $parameters An array of parameters for the message
+     * @param string|null $domain     The domain for the message or null to use the default
+     * @param string|null $locale     The locale or null to use the default
+     *
+     * @throws \InvalidArgumentException If the locale contains invalid characters
+     *
+     * @return string The translated string
+     *
+     * @api
+     */
+    public function transChoice($id, $number, array $parameters = array(), $domain = null, $locale = null)
+    {
+        // TODO: to be done when legacy will accept this (or will be replaced).
```

Comment:
Is this a wanted behavior ? Missing something here to avoid always throwing Exception I guess